app/services/search_indexer.rb

```diff
@@ -14,63 +14,109 @@ def self.scrub_html_for_search(html)
-                                    search_data: search_data,
-                                    id: id,
-                                    locale: SiteSetting.default_locale,
-                                    version: Search::INDEX_VERSION)
+                                    statement_params)
     if rows == 0
       Post.exec_sql("INSERT INTO #{table_name}
                     (#{foreign_key}, search_data, locale, raw_data, version)
-                    VALUES (:id, TO_TSVECTOR('#{stemmer}', :search_data), :locale, :raw_data, :version)",
-                                    raw_data: raw_data,
-                                    search_data: search_data,
-                                    id: id,
-                                    locale: SiteSetting.default_locale,
-                                    version: Search::INDEX_VERSION)
+                    VALUES (:id, #{ts_vector}, :locale, :raw_data, :version)",
+                                      statement_params)
     end
-  rescue
+
+    rescue
     # don't allow concurrency to mess up saving a post
   end
 
+  # for user login and name use "simple" lowercase stemmer
+  def self.stemmer_from_table(table)
+    table == "user" ? "simple" : Search.ts_config
+  end
+
+  # insert some extra words for I.am.a.word so "word" is tokenized
+  # I.am.a.word becomes I.am.a.word am a word
+  def self.prepare_entry_for_indexing(raw_data)
+    raw_data.gsub(/[^[:space:]]*[\.]+[^[:space:]]*/) do |with_dot|
+      split = with_dot.split(".")
+      if split.length > 1
+        with_dot + (" " << split[1..-1].join(" "))
+      else
+        with_dot
+      end
+    end
+  end
+
+  def self.params_for_update_index_statement(id, indexable_entries, raw_data)
+    statement_params = {
+      raw_data: raw_data,
+      id: id,
+      locale: SiteSetting.default_locale,
+      version: Search::INDEX_VERSION
+    }
+
+    indexable_entries.each.with_index do |indexable_entry, index|
+      statement_params[update_index_param_name_for(index)] = indexable_entry
+    end
+
+    statement_params
+  end
+
+  def self.update_index_param_name_for(index)
+    "indexable_fragment_#{index}".to_sym
+  end
+
+  def self.build_ts_vector(stemmer, indexable_entries)
+    raise ArgumentError, "The max number of entries to index is 4 to match the number of weights supported by PostgreSQL (A, B, C, D)" unless indexable_entries.length.between?(1, 4)
+
+    ts_vectors = indexable_entries.collect.with_index do |_, index|
+      "TO_TSVECTOR('#{stemmer}', :#{update_index_param_name_for(index)})"
+    end
+
+    if indexable_entries.length == 1
+      # index without weights
+      ts_vectors.first
+    else
+      weight_letters = ('A'..'D').to_a
+      ts_vectors.collect.with_index do |ts_vector, index|
+        weight_letter = weight_letters[index]
+        "setweight(#{ts_vector}, '#{weight_letter}')"
+      end.join(" || ")
+    end
+  end
+
   def self.update_topics_index(topic_id, title, cooked)
```

Comment:
While weights were only needed for posts, I added them for topics too for consistency